checker/src/visitors.rs

```diff
@@ -1345,152 +1368,153 @@ impl<'a, 'b: 'a, 'tcx: 'b, E> MirVisitor<'a, 'b, 'tcx, E> {
+                };
                 let span = self.current_span;
-                let warning = self.session.struct_span_warn(span, warning.as_str());
-                self.emit_diagnostic(warning);
-            }
 
-            // Now try to push a precondition so that any known or unknown caller of this function
-            // is warned that this function will fail if the precondition is not met.
-            if self.preconditions.len() < k_limits::MAX_INFERRED_PRECONDITIONS {
-                self.preconditions.push((
-                    self.current_environment
-                        .entry_condition
-                        .clone()
-                        .not(None)
-                        .or(cond.clone(), None)
-                        .replacing_provenance(self.current_span),
-                    warning,
-                ));
-            }
-            return;
-        }
-        if *known_name == KnownFunctionNames::StdBeginPanic {
-            let mut path_cond = self.current_environment.entry_condition.as_bool_if_known();
-            if path_cond.is_none() {
-                // Try the SMT solver
-                let path_expr = &self.current_environment.entry_condition.domain.expression;
-                let path_smt = self.smt_solver.get_as_smt_predicate(path_expr);
-                if self.smt_solver.solve_expression(&path_smt) == SmtResult::Unsatisfiable {
-                    path_cond = Some(false)
-                }
-            }
-            if !path_cond.unwrap_or(true) {
-                // We never get to this call, so nothing to report.
-                return;
-            }
-
-            let msg = if let Expression::CompileTimeConstant(ConstantDomain::Str(ref msg)) =
-                actual_args[0].1.domain.expression
-            {
-                if msg.contains("entered unreachable code") {
-                    // We treat unreachable!() as an assumption rather than an assertion to prove.
-                    return;
+                if path_cond.unwrap_or(false) && is_public(self.def_id, &self.tcx) {
+                    // We always get to this call and we have to assume that the function will
+                    // get called, so keep the message certain.
+                    let err = self.session.struct_span_warn(span, msg.as_str());
+                    self.emit_diagnostic(err);
                 } else {
-                    msg.clone()
+                    // We might get to this call, depending on the state at the call site.
+                    //
+                    // In the case when an assert macro has been called, the inverse of the assertion
+                    // was conjoined into the entry condition and this condition was simplified.
+                    // We therefore cannot distinguish between the case of maybe reaching a definitely
+                    // false assertion from the case of definitely reaching a maybe false assertion.
```

Comment:
Now I'm just picking on your grammar, but:

  "between" "from" -> "between" "and" OR  remove "between"